### ISSUE TITLE
print each matching element on one line

### ISSUE THREAD

**Issue by jackphel** [ericchiang/pup#49]
When I do `curl -s https://news.ycombinator.com/item?id=9996333 | pup '.comment'` I'm having a hard time taking the output and passing it along, maybe greping for text in the matched elements or sorting or something.
Do you see an easier way to use the output than having an option to put each matched element on one line?

**Comment by jackphel** [comment]
I ended up doing 
```
curl -s https://news.ycombinator.com/item?id=9996333 | pup '.comment' | awk '{if ($0 ~ /^<\span/) {print line; line="";} else line=line $0}'
```
after some sleuthing (because only the last outputted line of each comment element matches that regex), so I can pipe it forward like 
```
curl -s https://news.ycombinator.com/item?id=9996333 | pup '.comment' | awk '{if ($0 ~ /^<\span/) {print line; line="";} else line=line $0}' | grep 'esigner' | grep ' CA'
```

This would be super useful if I could just do 
```
curl -s https://news.ycombinator.com/item?id=9996333 | pup --compact-output '.comment'  | grep 'esigner' | grep ' CA'
```

like in `jq` 

```
 o   --compact-output / -c:

           By  default, jq pretty-prints JSON output. Using this option will result
           in more compact output by instead putting each JSON object on  a  single
           line.
```

**Comment by ericchiang** [comment]
Perhaps something like this?

```
curl -s https://news.ycombinator.com/item?id=9996333 | \
    pup '.comment p:contains("Python") a attr{href}'
```

**Comment by jackphel** [comment]
Thanks, that works for me!

**Comment by jackphel** [comment]
Thank you for your response. I think I was a little trigger happy in closing the question, though, because I'm still a little unsure about something. If I want to get the body of each comment to display, I can use `text{}`,
 e.g. 
```
curl -s https://news.ycombinator.com/item?id=9996333 | pup '.comment p:contains("Python") text{}'
```
 but then I'm still looking at lines of output rather than lines of matched elements so it seems difficult to do whatever with them, like putting each in its own file or putting them into lines of a report.

**Comment by ericchiang** [comment]
I can see what you're saying, that you need to be able to do some sort of 'foreach' mechanism.

I'll add something for this later.

For now, you can force each returned result into a single line using json and jq:

```
curl -s https://news.ycombinator.com/item?id=9996333 | \
    pup '.comment p:contains("Python") json{}' | \
    jq '.[]' -c 
```

**Comment by jackphel** [comment]
Thank you for your response. I think this project is awesome, by the way.
A foreach to something like `jq '.[] | {user: .user, title: .title}'` except with html. That would cut down so much scripting.

I hope you'll consider the `-c` flag, too, since it makes wiring up other unix tools super easy.

**Comment by tylerlong** [comment]
:+1:  I need this feature too

For a single query, I want to get an array of HTML text, instead of a merged HTML text.

**Comment by ericchiang** [comment]
@tylerlong I believe @jackphel wanted a newline for compliance with other cli tools. For example, to pipe into xargs.

**Comment by tylerlong** [comment]
@ericchiang Yes I agree with you. But I think in most cases `other cli tools` are using `\n` as delimiter. For example `grep` returns matching **lines**, so internally it splits the whole text by `\n`. So what I request is more universal.

**Comment by ericchiang** [comment]
@tylerlong could you give an example of when you'd need a more specific delimiter? How do you guarantee that that delimiter isn't printed as part of an HTML element?

**Comment by tylerlong** [comment]
@ericchiang 
I think in 99% cases '\n' as delimiter is enough. An example for the other 1%:  we want the splitted text to be formatted and indented HTML instead of a single line.
An real example is I wrote an Cocoa app which calls command line `pup` using `NSTask` and get standard output as `String`. Then in swift code I need to split the string to get an array. In this example though I am OK with `\n` as delimiter because I don't need indented HTML.

**Comment by ericchiang** [comment]
> > How do you guarantee that that delimiter isn't printed as part of an HTML element?

> He who picks the delimiter is responsible for that. It's not pup's concern.

The point I'm making is that you can't pick one without scanning the entire HTML at least once.

Just use an HTML parser written in Swift. That will fit your use case better.

**Comment by tylerlong** [comment]
@ericchiang 
It's very easy to pick a delimiter. example: `"#$%^&%^&*()*&^"`. providing I know where the HTML is from, it's easy to create a random delimiter. Just use very uncommon string.
Anyway, I have solved my problem already. Just my 2 cents on this, you don't have to take it serious.
Thank you for your reply and patience.

**Comment by jackphel** [comment]
@ericchiang, piping is exactly what I'd like to do :smile:. 

If matching elements get separated with newlines, it's difficult to determine which newlines were in the elements and which were added after. Perhaps html encoding the inner newlines could suffice to preserve them?

e.g. 
```
<div style="white-space: pre-wrap;">hey
buddy</div>
<div style="white-space: pre-wrap;">hey
buddy</div>
```
 -> `<div style="white-space: pre-wrap;">hey&#x0A;buddy</div>\n<div style="white-space: pre-wrap;">hey&#x0A;buddy</div>\n`

@tylerlong, perhaps using the proposed `-c` then find/replacing unescaped newlines with whatever delimiter accomplishes arbitrary delimiting?

**Comment by skojin** [comment]
+1 
I need single line output too, or at least option to not format output, so I can remove newlines from curl output with`tr -d "\n\r"`

**Comment by GrzegorzKazana** [comment]
If anyone is searching for a hack/solution, this worked well for me:

```bash
HTML_SPLIT_TOKEN="<!---SPLIT--->"
TRIM_REGEX='\S+.+\S+'
NEWLINE_CHARS="\n\r"

cat index.html |\
  pup 'some selectors' |\
  sed -Ee "s/^(<[a-z])/$HTML_SPLIT_TOKEN\1/g" |\ # insert split token before each unindented opening tag
  grep -oE "$TRIM_REGEX" |\                      # remove leading and trailing whitespaces in each line
  tr -d "$NEWLINE_CHARS" |\                      # remove all line breaks
  awk '$1=$1' RS="$HTML_SPLIT_TOKEN"             # split whole text by inserted token
```

tip:
you may need to escape exclamation mark in `<!---SPLIT--->` if you want to inline the variable
